### PR TITLE
util: remove the optimization in CopySelectedJoinRows() to fix a bug in OuterHashJoin

### PR DESCRIPTION
**Pull request by fzhedu**
### What problem does this PR solve? 
Fix https://github.com/pingcap/tidb/issues/13933

### What is changed and how it works?
Remove the optimization in CopySelectedJoinRows(), which tries to copy continuous outer rows to output rows, but it requests all outer rows are the same  as noted on the top of the function. But the request cannot be satisfied in OuterHashJoin. With respect to the performance, we create a new function `CopySelectedJoinRowsDirect`.

```
  GO111MODULE=on go test -bench=BenchmarkCopySelectedJoinRows
OK: 56 passed
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/util/chunk
BenchmarkCopySelectedJoinRows-12          	   54289	     21622 ns/op	       0 B/op	       0 allocs/op
BenchmarkCopySelectedJoinRowsDirect-12    	   29990	     39378 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/pingcap/tidb/util/chunk	3.273s
```
### Check List 

Tests 

 - unit test

Code changes

 - changed functions

Side effects

 - N/A

Related changes

 - N/A

Release note

 - N/A

### CODE REVIEW

**Comment by fzhedu** [comment]
The bug also occurs in indexHashJoin.
```
Create database fzh;
Create table s (a int, b int, key(b));
Create table t (a int, b int, key(b));
Insert into s values (1,2),(2,1),(11,11);
Insert into t values (11,2),(1,2),(5,2);
select /*+ INL_HASH_JOIN(s)*/ * from t left join s on s.b=t.b and s.a < t.a;
```
explain results:
```
mysql> explain  select /*+ INL_HASH_JOIN(s)*/ * from t left join s on s.b=t.b and s.a < t.a;
+-------------------------+----------+-----------+--------------------------------------------------------------------------------------------------------------+
| id                      | count    | task      | operator info                                                                                                |
+-------------------------+----------+-----------+--------------------------------------------------------------------------------------------------------------+
| IndexHashJoin_22        | 12475.01 | root      | left outer join, inner:IndexLookUp_11, outer key:fzh.t.b, inner key:fzh.s.b, other cond:lt(fzh.s.a, fzh.t.a) |
| ├─TableReader_24        | 10000.00 | root      | data:TableScan_23                                                                                            |
| │ └─TableScan_23        | 10000.00 | cop[tikv] | table:t, range:[-inf,+inf], keep order:false, stats:pseudo                                                   |
| └─IndexLookUp_11        | 1.25     | root      |                                                                                                              |
|   ├─Selection_9         | 1.25     | cop[tikv] | not(isnull(fzh.s.b))                                                                                         |
|   │ └─IndexScan_7       | 1.25     | cop[tikv] | table:s, index:b, range: decided by [eq(fzh.s.b, fzh.t.b)], keep order:false, stats:pseudo                   |
|   └─Selection_10        | 1.25     | cop[tikv] | not(isnull(fzh.s.a))                                                                                         |
|     └─TableScan_8       | 1.25     | cop[tikv] | table:s, keep order:false, stats:pseudo                                                                      |
+-------------------------+----------+-----------+--------------------------------------------------------------------------------------------------------------+
8 rows in set (0.01 sec)
```
wrong results from master:
```
mysql> select /*+ INL_HASH_JOIN(s)*/ * from t left join s on s.b=t.b and s.a < t.a;
+------+------+------+------+
| a    | b    | a    | b    |
+------+------+------+------+
|   11 |    2 |    1 |    2 |
|    1 |    2 |    1 |    2 |
|    1 |    2 | NULL | NULL |
+------+------+------+------+
3 rows in set (0.00 sec)
```
correct results from this PR:
```
mysql> select /*+ INL_HASH_JOIN(s)*/ * from t left join s on s.b=t.b and s.a < t.a;
+------+------+------+------+
| a    | b    | a    | b    |
+------+------+------+------+
|   11 |    2 |    1 |    2 |
|    5 |    2 |    1 |    2 |
|    1 |    2 | NULL | NULL |
+------+------+------+------+
3 rows in set (0.00 sec)
```

**Comment by fzhedu** [comment]
> Could you add some unit tests?

OK，make some cases where the outer parts of rows in a chunk are not the same.

File: util/chunk/chunk_util_test.go [hunk 1]
```diff
@@ -55,20 +67,45 @@ func TestCopySelectedJoinRows(t *testing.T) {
 		t.Fatal()
 	}
 }
+func TestCopySelectedJoinRowsDirect(t *testing.T) {
```
Comment (SunRunAway) [inline util/chunk/chunk_util_test.go]: Please add an empty line between functions.

File: util/chunk/chunk_util_test.go [hunk 2]
```diff
@@ -14,22 +14,34 @@
...
+	"math/rand"
 	"reflect"
 	"testing"
 
 	"github.com/pingcap/tidb/types"
 )
 
-func getChk() (*Chunk, *Chunk, []bool) {
+// getChk generate a chunk of data, isFirst3ColTheSame means the first three columns are the same.
+func getChk(isFirst3ColTheSame bool) (*Chunk, *Chunk, []bool) {
```
Comment (SunRunAway) [inline util/chunk/chunk_util_test.go]: isFirst3ColTheSame -> isLast3ColTheSame

File: util/chunk/chunk_util.go [hunk 3]
```diff
@@ -15,6 +15,51 @@ package chunk
 
 import "github.com/pingcap/errors"
 
+// CopySelectedJoinRowsDirect directly copies the selected joined rows from the source Chunk
+// to the destination Chunk.
+// Return true if at least one joined row was selected.
+func CopySelectedJoinRowsDirect(src *Chunk, innerColOffset, outerColOffset int, selected []bool, dst *Chunk) (bool, error) {
```
Comment (SunRunAway) [inline util/chunk/chunk_util.go]: `innerColOffset` and `outerColOffset` seems to be useless.

File: util/chunk/chunk_util_test.go (hunk 1 above)
Comment (fzhedu) [reply util/chunk/chunk_util_test.go]: done

File: util/chunk/chunk_util_test.go (hunk 2 above)
Comment (fzhedu) [reply util/chunk/chunk_util_test.go]: done

File: util/chunk/chunk_util.go (hunk 3 above)
Comment (fzhedu) [reply util/chunk/chunk_util.go]: removed.

File: executor/join_test.go [hunk 4]
```diff
@@ -1771,6 +1771,24 @@ func (s *testSuiteJoin1) TestIssue11390(c *C) {
...
 
+func (s *testSuiteJoin1) TestOuterTableBuildHashTable(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t, s")
+	tk.MustExec("create table t (a int,b int)")
+	tk.MustExec("create table s (a int,b int)")
+	tk.MustExec("insert into t values (11,11),(1,2)")
+	tk.MustExec("insert into s values (1,2),(2,1),(11,11)")
+	tk.MustQuery("select * from t left join s on s.a > t.a").Sort().Check(testkit.Rows("1 2 11 11", "1 2 2 1", "11 11 <nil> <nil>"))
```
Comment (SunRunAway) [inline executor/join_test.go]: We should add outer join hint in PR #13826

Comment (fzhedu) [reply executor/join_test.go]: The current optimizer can choose the outer hash join. We should add the hint after merge the PR.

**Comment by SunRunAway** [comment]
@qw4990 @XuHuaiyu PTAL

File: executor/join_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (fzhedu) [reply executor/join_test.go]: if without this PR, the explain results are also correct. The most important is to let the SQL run the `trytojoinOuter` by using hints.

File: executor/joiner.go [hunk 5]
```diff
@@ -222,12 +222,7 @@ func (j *baseJoiner) filterAndCheckOuterRowStatus(input, output *chunk.Chunk, in
 	}
 
 	// Batch copies selected rows to output chunk.
-	innerColOffset, outerColOffset := 0, innerColsLen
-	if !j.outerIsRight {
-		innerColOffset, outerColOffset = input.NumCols()-innerColsLen, 0
-	}
-
-	_, err = chunk.CopySelectedJoinRows(input, innerColOffset, outerColOffset, j.selected, output)
+	_, err = chunk.CopySelectedJoinRowsDirect(input, j.selected, output)
```
Comment (SunRunAway) [inline executor/joiner.go]: Does each inner row is the same here?
Could we implement and use  `CopySelectedJoinRowsWithSameInnerRows` here?

Comment (fzhedu) [reply executor/joiner.go]: Yes, but in practice, the improvement is rare in a query's execution, even though the benchmark on small size data is better. This is like the benefits from vectorized expression.

**Pull request merged**